### ISSUE TITLE
Allow for async actions before API requests

### ISSUE THREAD

**Issue by bendavis78** [swagger-api/swagger-js#701]
Use case: it's good to check for for expired tokens before every API request, and refreshing tokens is an asynchronous operation. Right now I have to have this before every api call:

    checkExpired().then(function() {
        return client.default.createSomething(params);
    });

It would be ideal if I could intercept requests asynchronously.

**Comment by fehguy** [comment]
I believe the requestInterceptor will do the trick. Take a look at the client.js test for usage examples

**Comment by bendavis78** [comment]
I tried `requestInterceptor`, but `SwaggerHttp` [doesn't wait](https://github.com/swagger-api/swagger-js/blob/4aab2557a6625cbf800f05d0f121f8cc56817ba7/lib/http.js#L92) for anything before sending the request.

The object being passed around has a `deferred` property, but that's a promise for the response. I need the client to wait on a promise for something that's needed before the request.
